### PR TITLE
Add functionality to utilise the .forceignore file for exclusions  

### PR DESCRIPTION
**Pull request by jjulicher**
## What does this pull request contains? Explain your changes.
Add functionality to utilise the .forceignore file for exclusions  as lwc directory can contain files that are not to be included in the package.xml. Includes additional CLI parameter -i / --ignore to specify file .
---

- [x] Added for new features.
- [ ] Changed for changes in existing functionality.
- [ ] Deprecated for soon-to-be removed features.
- [ ] Removed for now removed features.
- [ ] Fixed for any bug fixes.
- [ ] Security in case of vulnerabilities.

## Explain your changes

---

Contains additional config & tests to filter using ignore functionality for the repoGitDiff util. 
Added in command line parameters to specify the ignore file. 

## Does this close any currently open issues?

---

no

- [ ] Jest test to check the fix is applied are added.

## Any particular element to being able to test locally

---

_(Write your answer here.)_

## Any other comments?

---

_(Write your answer here.)_

## Where has this been tested?

---

**Operating System:** …
Windows/WSL 
**NPM version:** … 
6.14.8
**Node version:** …
14.9.0
**sgd version:** …
3.3.1
**git version:** …
2.17.1

### CODE REVIEW

File: lib/utils/repoGitDiff.js
Comment (scolladon) [inline lib/utils/repoGitDiff.js:17]: Remove the unused space

File: lib/utils/repoGitDiff.js
Comment (scolladon) [inline lib/utils/repoGitDiff.js:7]: Remove ; at the end of the line

File: lib/utils/repoGitDiff.js
Comment (scolladon) [inline lib/utils/repoGitDiff.js:8]: Order the package import alphabetically : 
cpUtil
gc
fs
os
path
ignore

File: lib/utils/repoGitDiff.js
Comment (scolladon) [inline lib/utils/repoGitDiff.js:30]: `
.filter(line => !ig.ignores(line))
`
Could work ?

File: __tests__/unit/lib/utils/repoGitDiff.test.js
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:6]: Is it really working without this mock ?
If yes just remove the line plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:15]: use a constant at the beginning of the file for the path to the forceignore file plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:28]: use a constant at the beginning of the file for the path to the forceignore file plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:64]: use a constant at the beginning of the file for the path to the forceignore file plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:46]: Leave the ',' at the end of the line
ecmascript support that

File: __tests__/unit/lib/utils/repoGitDiff.test.js
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:34]: Leave the ',' at the end of the line
ecmascript support that

File: __tests__/unit/lib/utils/.forceignore
Comment (scolladon) [inline __tests__/unit/lib/utils/.forceignore:10]: I would prefer to move it under the __mocks__ folder 
(and adjust code reference to it in the repoGitDiff.test.js file)

**Review by scolladon — CHANGES_REQUESTED** [review]
Wonderfull suggestion, I really like it.

Please @jjulicher consider this review

And also could you add a test case for the ignore in the cli.test.js test file please

I am open to discussion :)

File: README.md
Comment (scolladon) [inline README.md:51]: Add also a reference to the ignore module in the Built With section plz

File: lib/utils/repoGitDiff.js
Comment (jjulicher) [reply lib/utils/repoGitDiff.js:30]: Unfortunately line was sometimes null/undefined and ignores would throw an error.

File: __tests__/unit/lib/utils/repoGitDiff.test.js
Comment (jjulicher) [reply __tests__/unit/lib/utils/repoGitDiff.test.js]: Yes it did! 

Ran all tests & passed - needed it to test ignore loading a file properly. Didn't seem to cause any other issues 

will remove.

**Review by scolladon — CHANGES_REQUESTED** [review]
Please don't remove .gitignore (just don't change it)
use npm run lint:fix to make sure everything is compliant with the coding style
And review the other comments please

**Comment by scolladon** [comment]
In order to resolve the conflicts please rebase the branch on top of origin/master

File: lib/utils/repoGitDiff.js
Comment (scolladon) [reply lib/utils/repoGitDiff.js:31]: So we could rewrite it like this : 
`
.filter(line => line && !ig.ignores(line))
`
?

File: lib/utils/repoGitDiff.js
Comment (scolladon) [inline lib/utils/repoGitDiff.js:10]: Linting issue here

File: lib/utils/repoGitDiff.js
Comment (jjulicher) [reply lib/utils/repoGitDiff.js:31]: ahh yes of course. that should work.  I'll update.

File: README.md
Comment (jjulicher) [reply README.md:51]: Done.

File: lib/utils/repoGitDiff.js
Comment (scolladon) [inline lib/utils/repoGitDiff.js:31]: in fact we should use [double negation](https://riptutorial.com/javascript/example/3047/double-negation----x-) check here for the code to be more robust.

Could you change it to 
`
.filter(line => !!line && !ig.ignores(line))
`

**Comment by scolladon** [comment]
We may have an issue here...
It seems commit from other branch have been added here...

**Comment by jjulicher** [comment]
Oh crap. Sorry it's late after a long week.... origin/master is from my
repo which was already merged with all the changes for my CI pipeline.

I'm going to have to create a new branch aren't I... 🙄😪 And reintegrate
all my changes.....

**Comment by scolladon** [comment]
No worries, 

I am rebasing the work right now and I am applying the fixes required.
I'll let you have a look before merging.

Cheers

**Review by scolladon — DISMISSED** [review]

**Comment by scolladon** [comment]
@jjulicher everything looks ok for me, is it ok for you too ?

@mehdisfdc are you ok to merge this feature ?

**Comment by jjulicher** [comment]
> No worries,
> 
> I am rebasing the work right now and I am applying the fixes required.
> I'll let you have a look before merging.
> 
> Cheers

Oh thanks!

**Comment by jjulicher** [comment]
> @jjulicher everything looks ok for me, is it ok for you too ?
> 
> @mehdisfdc are you ok to merge this feature ?

Looks good to me.

**Review by scolladon — APPROVED** [review]

**Comment by mehdisfdc** [comment]
> @jjulicher everything looks ok for me, is it ok for you too ?
> 
> @mehdisfdc are you ok to merge this feature ?

👍 

Thank you @jjulicher for your contribution!

**Pull request merged**
